src/jvmMain/kotlin/io/beatmaps/api/users.kt

```diff
@@ -698,6 +761,51 @@ fun Route.userRoute() {
 
         call.respond(UserDetail.from(user, stats = statsForUser(user)))
     }
+
+    fun getFollowerData(page: Long, joinOn: Column<EntityID<Int>>, condition: SqlExpressionBuilder.() -> Op<Boolean>) = transaction {
```

Comment:
Can you not remove a whole join here?

```kotlin
    fun getFollowerData(page: Long, joinOn: Column<EntityID<Int>>, condition: SqlExpressionBuilder.() -> Op<Boolean>) = transaction {
        User
            .join(Beatmap, JoinType.LEFT, User.id, Beatmap.uploader) {
                Beatmap.deletedAt.isNull()
            }
            .slice(
                User.id, User.name, User.uniqueName, User.description, User.avatar, User.discordId, User.hash,
                Beatmap.id.count(), User.admin, User.curator, User.verifiedMapper
            )
            .select {
                User.id.inSubQuery(
                    Follows
                        .slice(joinOn)
                        .select(condition)
                        .limit(20, page)
                        .orderBy(Follows.since, SortOrder.DESC)
                )
            }
            .groupBy(User.id)
            .map { row ->
                UserDetail.from(UserDao.wrapRow(row), stats = UserStats(totalMaps = row[Beatmap.id.count()].toInt()))
            }
    }
```

Reply:
I was sort of wrong here, we need the join so we can resort on since at the outer level

Reply:
More code incoming. This does perform consistently slightly better than the existing code. Uses a subquery in select which allows us to get the since column out rather than the in above which doesn't.

```kotlin
    fun getFollowerData(page: Long, joinOn: Column<EntityID<Int>>, condition: SqlExpressionBuilder.() -> Op<Boolean>) = transaction {
        val followsSubquery = Follows
            .slice(joinOn, Follows.since)
            .select(condition)
            .limit(page)
            .orderBy(Follows.since, SortOrder.DESC)
            .alias("fs")

        followsSubquery
            .join(User, JoinType.LEFT, followsSubquery[joinOn], User.id)
            .join(Beatmap, JoinType.LEFT, User.id, Beatmap.uploader) {
                Beatmap.deletedAt.isNull()
            }
            .slice(
                User.id, User.name, User.uniqueName, User.description, User.avatar, User.discordId, User.hash,
                Beatmap.id.count(), User.admin, User.curator, User.verifiedMapper
            )
            .selectAll()
            .groupBy(User.id, followsSubquery[Follows.since])
            .orderBy(followsSubquery[Follows.since], SortOrder.DESC)
            .map { row ->
                UserDetail.from(UserDao.wrapRow(row), stats = UserStats(totalMaps = row[Beatmap.id.count()].toInt()))
            }
    }
```